packages/keyring/src/keyring.ts

```diff
@@ -0,0 +1,333 @@
+    skipAddressIndexes = []
+  }: GenerateAddressProps): SensitiveAddressData => {
+    if (group !== undefined && (!Number.isInteger(group) || group < 0 || group >= TOTAL_NUMBER_OF_GROUPS))
+      throw new Error(`Keyring: Could not generate address in group ${group}, group is invalid`)
+
+    let nextAddressIndex = skipAddressIndexes.includes(addressIndex)
+      ? findNextAvailableAddressIndex(addressIndex, skipAddressIndexes)
+      : addressIndex
+    let newAddressData = this._deriveAddressAndKeys(nextAddressIndex)
+
+    while (group !== undefined && addressToGroup(newAddressData.hash, TOTAL_NUMBER_OF_GROUPS) !== group) {
+      nextAddressIndex = findNextAvailableAddressIndex(newAddressData.index, skipAddressIndexes)
+      newAddressData = this._deriveAddressAndKeys(nextAddressIndex)
+    }
+
+    return newAddressData
+  }
+
+  private _deriveAddressesInGroup = (group: number, amount: number, skipIndexes: number[]): SensitiveAddressData[] => {
+    const addresses = []
+    const skipAddressIndexes = Array.from(skipIndexes)
+
+    for (let j = 0; j < amount; j++) {
+      const newAddress = this._generateAddress({ group, skipAddressIndexes })
+      addresses.push(newAddress)
+      skipAddressIndexes.push(newAddress.index)
+    }
+
+    return addresses
+  }
+
+  private _getNonSensitiveAddressData = ({
+    hash,
+    index,
+    publicKey
+  }: SensitiveAddressData): NonSensitiveAddressData => ({ hash, index, publicKey })
+
+  private _initFromMnemonic = (mnemonic: Uint8Array | null, passphrase: string) => {
+    if (this.root) throw new Error('Keyring: Secret recovery phrase already provided')
+    if (!mnemonic) throw new Error('Keyring: Secret recovery phrase not provided')
+
+    const isValid = validateMnemonic(mnemonic, wordlist)
+    if (!isValid) throw new Error('Keyring: Invalid secret recovery phrase provided')
+
+    const seed = mnemonicToSeedSync(mnemonic, wordlist, passphrase)
+    this.hdWallet = HDKey.fromMasterSeed(seed)
+    this.root = this.hdWallet.derive(this.hdPath)
+
+    passphrase = ''
+    mnemonic = null
+  }
+
+  private _deriveAddressAndKeys = (addressIndex: number): SensitiveAddressData => {
+    if (!this.root) throw new Error('Keyring: Cannot derive address and keys, secret recovery phrase is not provided')
+    if (!isAddressIndexValid(addressIndex)) throw new Error('Invalid address index path level')
+
+    const keyPair = this.root.deriveChild(addressIndex)
+
+    if (!keyPair.publicKey) throw new Error('Keyring: Missing public key')
+    if (!keyPair.privateKey) throw new Error('Keyring: Missing private key')
+
+    const publicKey = bytesToHex(keyPair.publicKey)
+    const privateKey = keyPair.privateKey
+    const hash = blake.blake2b(Uint8Array.from(keyPair.publicKey), undefined, 32)
+    const type = new Uint8Array([0])
+    const bytes = new Uint8Array(type.length + hash.length)
+    bytes.set(type, 0)
+    bytes.set(hash, type.length)
+    const address = bs58.encode(bytes)
```

Comment:
This is an important part, we used to use `Buffer`, but to make it work with the bip packages that Metamask uses it had to (rightfully) be changed to `Uint8Array`.